test/e2e/native_histograms_test.go

```diff
@@ -96,48 +110,263 @@ func TestWriteNativeHistograms(t *testing.T) {
+	ts := func() time.Time { return now }
+
+	queryAndAssert(t, ctx, querier.Endpoint("http"), func() string { return testHistogramMetricName }, ts, promclient.QueryOptions{Deduplicate: true}, expectedHistogramModelVector(histograms[0], map[string]string{
 		"tenant_id": "default-tenant",
 	}))
 }
 
-func synthesizeHistogram(ctx context.Context, rawRemoteWriteURL string) error {
+func TestQueryFrontendNativeHistograms(t *testing.T) {
+	e, err := e2e.NewDockerEnvironment("nat-hist-qfe")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	prom1, sidecar1 := e2ethanos.NewPrometheusWithSidecar(e, "ha1", e2ethanos.DefaultPromConfig("prom-ha", 0, "", "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "", "native-histograms", "remote-write-receiver")
+	prom2, sidecar2 := e2ethanos.NewPrometheusWithSidecar(e, "ha2", e2ethanos.DefaultPromConfig("prom-ha", 1, "", "", e2ethanos.LocalPrometheusTarget), "", e2ethanos.DefaultPrometheusImage(), "", "native-histograms", "remote-write-receiver")
+	testutil.Ok(t, e2e.StartAndWaitReady(prom1, sidecar1, prom2, sidecar2))
+
+	querier := e2ethanos.NewQuerierBuilder(e, "querier", sidecar1.InternalEndpoint("grpc"), sidecar2.InternalEndpoint("grpc")).Init()
+	testutil.Ok(t, e2e.StartAndWaitReady(querier))
+
+	inMemoryCacheConfig := queryfrontend.CacheProviderConfig{
+		Type: queryfrontend.INMEMORY,
+		Config: queryfrontend.InMemoryResponseCacheConfig{
+			MaxSizeItems: 1000,
+			Validity:     time.Hour,
+		},
+	}
+
+	queryFrontend := e2ethanos.NewQueryFrontend(e, "query-frontend", "http://"+querier.InternalEndpoint("http"), queryfrontend.Config{}, inMemoryCacheConfig)
+	testutil.Ok(t, e2e.StartAndWaitReady(queryFrontend))
+
+	rawRemoteWriteURL1 := "http://" + prom1.Endpoint("http") + "/api/v1/write"
+	rawRemoteWriteURL2 := "http://" + prom2.Endpoint("http") + "/api/v1/write"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	t.Cleanup(cancel)
+
+	histograms := tsdb.GenerateTestHistograms(4)
+	now := time.Now()
+	_, err = writeHistograms(ctx, now, histograms, rawRemoteWriteURL1)
+	testutil.Ok(t, err)
+	startTime, err := writeHistograms(ctx, now, histograms, rawRemoteWriteURL2)
+	testutil.Ok(t, err)
+
+	// Ensure we can get the result from Querier first so that it
+	// doesn't need to retry when we send queries to the frontend later.
+	queryAndAssertSeries(t, ctx, querier.Endpoint("http"), func() string { return testHistogramMetricName }, time.Now, promclient.QueryOptions{Deduplicate: true}, []model.Metric{
+		{
+			"__name__":   testHistogramMetricName,
+			"prometheus": "prom-ha",
+			"foo":        "bar",
+		},
+	})
+
+	vals, err := querier.SumMetrics([]string{"http_requests_total"})
+	e2emon.WithLabelMatchers(matchers.MustNewMatcher(matchers.MatchEqual, "handler", "query"))
```

Comment:
This value is unused? :thinking:

Reply:
You are tight, should be argument and I fixed it.